### DISCUSSION TITLE
Introduce a WebGLVectorLayer class to replace WebGLPointsLayer

### DISCUSSION THREAD

**Discussion by jahow** [openlayers/openlayers#16187] [Ideas]
The `WebGLPointsLayer` class has been around for a while and seems to be used quite a bit. We have been working on the more generic vector layer renderer for a while, but there hasn't been any proper easy-to-use class yet for that renderer.

Now that the generic vector renderer is working well, we propose to:

- Deprecate the WebGLPointsLayer class: it won't be maintained and will be removed in the n+1 major version; it should work moslty the same until then
- Introduce the WebGLVectorLayer class
- Introduce a way to specify style variables through this class API

The usage of the WebGLVectorLayer class would look like this:

```js
const layer = new WebGLVectorLayer({
	source: …,
	style: [{
		'stroke-color': ['var', 'lineColor']
	}],
	styleVariables: {
		lineColor: 'yellow'
	}
})
```

The `['var']` operator will still be supported but target the `styleVariables` property instad of the `variables` property in the style object. The layer API will also offer an `updateStyleVariables` function, similarly to the WebGLPointsLayer class.

This addresses at least partly https://github.com/openlayers/openlayers/issues/15146.

This proposal should result in a more easy-to-use and optimized WebGL-based layer. The generic vector layer is also better optimized and should offer better performance.

This proposal will also pave the way for the following improvements:
* implement style variables and `['var']` in a similar way for canvas-based layers
* get rid of the webgl-specific style format and use the flat style format everywhere
* implement cascading style variables on the Map and LayerGroup objects (to be discussed further)

Please share any opinion/questions!

Done in cooperation with @romprokop

**Comment by cugarteblair** [comment]
I welcome this.
I have been using openlayers with deck.gl PointCloudLayer to render large sets of points. 
However the cursor is stuck on "grabbing" and I can't seem to get openlayers and deck to play well together vis-a-vis mouse events.
I don't want to use anything that isn't in the official release so that is why I haven't tried WebGLPointsLayer because the example says it is not in an official release.
Is there an expected release for either of these features (WebGlPoints or this new one)?

**Reply by jahow** [reply]
I wouldn't give an estimate for when these classes will get out of the experimental status. There's still quite a way to go and unfortunately we've been operating on a tight funding lately. I'll try and establish a roadmap for the missing features so we can have a shared understanding of what's left to do!

**Reply by jahow** [reply]
Thank you for your interest!

**Reply by cugarteblair** [reply]
If there is any way I could contribute for that feature let me know

**Comment by jahow** [comment]
PR is up! https://github.com/openlayers/openlayers/pull/16394
